### PR TITLE
remove remote address book if access was revoked

### PR DESCRIPTION
**Pull request by schiessle**
Following scenario:
1. server1 adds server2 as trusted server
2. server2 adds server1 as trusted server
3. both exchange their system address books
4. admin of server2 removes server1 from the list of trusted servers

Current behavior:

both will still have the latest version of the synced address books and use it for auto-complete

Behavior with this PR:

both server will remove the synced address books. The status indicator on server1 will go back to yellow.

cc @DeepDiver1975 please have a look. Thanks!

### CODE REVIEW

**Comment by schiesbn** [comment]
@Helios07 this solves the problem we talked about.

**Comment by Helios07** [comment]
That's great! Thank you.

**Comment by karlitschek** [comment]
:+1:

File: apps/federation/appinfo/application.php [hunk 1]
```diff
@@ -94,6 +97,17 @@ private function registerService() {
 				$c->query('TrustedServers')
 			);
 		});
+
+		$container->registerService('CardDavBackend', function($c) {
```
Comment (DeepDiver1975) [inline apps/federation/appinfo/application.php]: we should not mix app code like this ... let me think on how to do this ....

Comment (DeepDiver1975) [inline apps/federation/appinfo/application.php]: @schiesbn please use the event dispatcher to communicate with the dav app - THX

Comment (schiesbn) [inline apps/federation/appinfo/application.php]: @DeepDiver1975 done... please have a look. Thanks!

**Comment by nickvergessen** [comment]
Looks good, but would be nice if we can find a place where we put the event names in one place, so app authors can 1. see which events exist (prevent duplicates) 2. know the exact casing/spelling of the names 3. make it kind of visible that event names are public api.

But that is a different story, so :+1:  and I will create a follow up issue for the name thing.

File: apps/federation/lib/trustedservers.php [hunk 2]
```diff
@@ -154,7 +164,10 @@ public function addSharedSecret($url, $sharedSecret) {
 	 * @param int $id
 	 */
 	public function removeServer($id) {
+		$server = $this->dbHandler->getServerById($id);
 		$this->dbHandler->removeServer($id);
+		$event = new GenericEvent('OCP\Federation\TrustedServerEvent::remove', $server);
```
Comment (DeepDiver1975) [inline apps/federation/lib/trustedservers.php]: first arg can be null 
second arg has to be an array

Comment (schiesbn) [inline apps/federation/lib/trustedservers.php]: I use now the subject do pass the url_hash directly... think that is what you meant... Right?

File: apps/federation/lib/trustedservers.php [hunk 3]
```diff
@@ -154,7 +164,10 @@ public function addSharedSecret($url, $sharedSecret) {
 	 * @param int $id
 	 */
 	public function removeServer($id) {
+		$server = $this->dbHandler->getServerById($id);
 		$this->dbHandler->removeServer($id);
+		$event = new GenericEvent($server['url_hash']);
```
Comment (nickvergessen) [inline apps/federation/lib/trustedservers.php]: I think the other way made more sense

**Comment by schiesbn** [comment]
@DeepDiver1975 final review please... Thanks!

File: apps/federation/lib/trustedservers.php (hunk 3 above)
Comment (schiesbn) [inline apps/federation/lib/trustedservers.php]: I think you can argue in both ways. This way the dav app doesn't need to have any "magic" knowledge about the server array. It just gets the subject which is the address book id, done.

Comment (nickvergessen) [inline apps/federation/lib/trustedservers.php]: yes, fine by me

File: apps/federation/appinfo/database.xml [hunk 4]
```diff
@@ -27,8 +27,7 @@
 				<type>text</type>
 				<default></default>
 				<notnull>true</notnull>
-				<length>32</length>
```
Comment (DeepDiver1975) [inline apps/federation/appinfo/database.xml]: change version because of schema change?
no length?

Comment (nickvergessen) [inline apps/federation/appinfo/database.xml]: It was changed, but concurrency of pull requests killed the change.

Comment (schiesbn) [inline apps/federation/appinfo/database.xml]: I increased the version with my last commit.

**Comment by schiesbn** [comment]
All tests passed, beside some unrelated swift test... @DeepDiver1975 can we get this in? Thanks!

**Comment by DeepDiver1975** [comment]
:+1:

**Pull request merged**